Library/Core/UnoCore/Source/Uno/Threading/Thread.uno

```diff
@@ -0,0 +1,183 @@
+using Uno;
+using Uno.Collections;
+using Uno.Compiler.ExportTargetInterop;
+using System.Globalization;
+
+namespace Uno.Threading
+{
+    [TargetSpecificType]
+    [extern(UNIX) Set("Include", "pthread.h")]
+    [extern(UNIX) Set("TypeName", "pthread_t")]
+    [extern(WIN32) Set("Include", "Uno/WinAPIHelper.h")]
+    [extern(WIN32) Set("TypeName", "HANDLE")]
+    extern(CPLUSPLUS) struct ThreadHandle
+    {
+    }
+
+    [TargetSpecificType]
+    [Set("Include", "Uno/Support.h")]
+    [Set("TypeName", "uThreadLocal*")]
+    extern(CPLUSPLUS) struct ThreadLocal
+    {
+    }
+
+    [DotNetType("System.Threading.ThreadStart")]
+    public delegate void ThreadStart();
+
+    [DotNetType("System.Threading.ThreadStateException")]
+    public class ThreadStateException : Exception
+    {
+        public ThreadStateException(String message) : base(message)
+        {
+        }
+    }
+
+    [DotNetType("System.Threading.Thread")]
+    [extern(UNIX) Require("Source.Declaration", "static void* _ThreadFunc(void* arg) { @{ThreadMain(Thread):Call((@{Thread}) arg)}; return NULL; }")]
+    [extern(WIN32) Require("Source.Declaration", "static DWORD WINAPI _ThreadFunc(LPVOID lpParam) { @{ThreadMain(Thread):Call((@{Thread}) lpParam)}; return 0; }")]
+    public sealed class Thread
+    {
+        extern(CPLUSPLUS) static ThreadLocal _currentThread;
+
+        extern(CPLUSPLUS) static Thread()
+        {
+            _currentThread = extern<ThreadLocal> "uCreateThreadLocal(NULL)";
+        }
+
+        extern(CPLUSPLUS) static void ThreadMain(Thread thread)
+        {
+            extern "uAutoReleasePool pool";
+            extern "uSetThreadLocal(@{_currentThread}, $0)";
+
+            try
+            {
+                thread._threadStart();
+            }
+            catch (Exception e)
+            {
+                // TODO: Use some kind of exception callback..
+                debug_log "Unhandled exception in thread: " + e;
```

Comment:
Previously, this would cause an unhandled C++ exception on the running thread (and leak the Thread-object).